@AuthenticationPrincipal을 통한 로그인 회원 정보 조회 기능 구현

kimcno3 opened the pull request:
* UserDetails의 User 클래스의 상속 클래스 생성(UserAdapter)
* UserAdapter 타입 객체 생성 메소드 구현
* CustomUserDetailsService.loadUserByUsername() 메소드 수정
  - 리턴 객체의 타입을 UserAdapter로 바꿔 user 객체를 가지고 있을 수 있도록 변경
* 컨트롤러 파라미터에서 user정보를 조회할 수 있는 리졸버 클래스 생성
* 생성한 리졸버 클래스를 리졸버 리스트에 추가하기 위한 configuration 클래스 생성

kimcno3 committed:
@AuthorizationPrincipal을 통한 로그인 회원 정보 조회 기능 구현

* UserDetails의 User 클래스의 상속 클래스 생성(UserAdapter)
* UserAdapter 타입 객체 생성 메소드 구현
* CustomUserDetailsService.loadUserByUsername() 메소드 수정
  - 리턴 객체의 타입을 UserAdapter로 바꿔 user 객체를 가지고 있을 수 있도록 변경
* 컨트롤러 파라미터에서 user정보를 조회할 수 있는 리졸버 클래스 생성
* 생성한 리졸버 클래스를 리졸버 리스트에 추가하기 위한 configuration 클래스 생성

kimcno3 commented:
`@AuthenticationPrincipal ` 을 활용해 로그인 회원 정보를 호출해보기 위해 어뎁터 패턴을 활용해보려고 했습니다!

문제는 계속해서 호출된 객체가 null 값을 가진다는 것이였고 저와 같은 문제를 가진 분이 작성한 블로그를 참고해봤습니다!
https://sillutt.tistory.com/entry/Spring-Security-AuthenticationPrincipal%EA%B3%BC-ArgumentResolver

스프링 부트 2.4 이후부터는 `@AuthenticationPrincipal `를 사용할 때 NPE가 발생하는 문제가 있고 블로그 작성자분께서 발견한 에러 이유는`@AuthenticationPrincipal `에 의해 동작하는 resolver의 검증 과정을 어뎁터 객체가 통과하지 못해 null값으로 리턴되는 것이라 했고, 이에 대한 해결 방법을 결국 커스텀 리졸버를 구현함으로써 해결했습니다!

일단 저도 의도한 동작이 하도록 커스텀 리졸버까지 구현했지만 처음 의도한대로 Spring Security에서 제공해주는 간편한 기능을 활용하지 못하는 점이 아쉽습니다ㅠㅠ

처음 의도한 대로`@AuthenticationPrincipal `를 활용해 정상적으로 동작시켜볼 수 있는 방법을 찾아볼 수 있는 키워드가 있을지 궁금합니다!

f-lab-bright commented:
음.. spring security에서 @AuthenticationPrincipal이 어떤 값을 어디서 가져오는지 확인해보시는게 필요할 것 같습니다 :)
null이 전달된다는건, security에서 의도한 객체나 저장소에 제대로 값이 주입되지 않았다는 이야기입니다 :)

kimcno3 committed:
주석문 추가

* resolver 객체 주입 코드 주석화
* 문제점 설명을 위한 주석문 추가

kimcno3 commented on src/main/java/com/flab/football/service/security/CustomUserDetailsService.java:
```diff
@@ -22,22 +25,56 @@ public class CustomUserDetailsService implements UserDetailsService {
   @Override
   public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
제가 알아본 바로는 `@AthenticationPrincipal` 로 가져오는 객체는 UserDetailsService.loadUserByUsername()의 리턴 객체로 알고 있습니다!!

그래서 이 메소드의 리턴 객체를 UserDetails 인터페이스 구현 클래스인 User 클래스를 상속한 UserAdapter 타입의 객체로 리턴하도록 수정해봤지만 컨트롤러에서 UserAdapter 클래스 타입의 객체로 선언하면 null값이 나오는 문제가 있습니다ㅠㅠ

Adapter 타입의 객체로 가져와 Entity 클래스인 User 객체를 활용해보고자 하고 있지만 같은 위 문제의 이유를 잡지 못하고 있는 상태입니다...ㅠㅠ

kimcno3 committed:
User 정보 조회 기능 리팩토링

* LoginUser 어노테이션 생성
* 커스텀 리졸버 빈 객체 등록
* LogInUserArgumentResolver로 커스텀 리졸버 클래스명 변경

kimcno3 committed:
User 정보 조회 기능 리팩토링

* userId 만 리턴하도록 리졸버 로직 변경
* 어노테이션, 리졸버 클래스명 변경
* ChatController에 LogInUserId 어노테이션 적용

kimcno3 commented:
UserId만을 리턴하도록 리졸버 로직을 변경했습니다!

f-lab-bright commented:
여기는 resolver 말고, @AuthenticationPrincipal을 쓸 수 있도록 한 번 해결책을 찾아보시면 좋을 것 같습니다 :)
jwt를 쓰는 환경에서는 디비 접근 전혀 없이, jwt의 payload로 로그인 된 사용자의 정보를 받아올 수 있도록 구성하는게 가장 좋습니다 :)

kimcno3 committed:
Jwt 토큰 payload 정보 양식 변경

* subject에 userId를 저장
  - 이에 따라 Athentication 타입 객체의 username 값은 userId로 변경

* SecurityUtil.getCurrentUserEmail() 메소드명 변경
 -> SecurityUtil.getCurrentUserId()로 변경

* JwtSecurityService의 불필요한 메소드 삭제
  - userId를 "sub"에 저장함으로써 userId를 별도로 저장할 소요가 없어졌다.
  - Claims의 특정 키를 활용해 가져 탐색하지 않고 getSubject() 로 userId 조회 가능

* @LogInUserId -> @AuthenticationPrincipal로 변경

kimcno3 commented:
Jwt 토큰에 저장되는 "sub" 값에 userEmail이 아닌 userId를 저장되도록 수정했습니다!

이를 통해 controller에서도 userId에 대한 접근을 UserDetails 타입의 객체로도 조회할 수 있도록 수정되었고 @AuthenticationPrincipal 어노테이션도 활용할 수 있도록 수정했습니다!

그리고 Spring Security의 인증, 인가 처리가 어떤 방식으로 동작하는지도 함께 이해 중이고 블로그에도 정리볼 생각입니다😁

kimcno3 committed:
WebSocket 서버에서 토큰을 해석해 userId를 가져오는 메소드에 대한 주석문 추가

f-lab-bright reviewed (approved):
LGTM with Comments
꼭 UserDetails 객체를 사용하지 않으셔도 됩니다.
내부적으로 인증된 유저를 관리하는 dto를 전달하고 받으시는게 좀 더 사용하기 편하실거에요 :)
